File: 프로그래머스/lv2/92342. 양궁대회/양궁대회.py
from copy import deepcopy

ans, apeach_arrows = [0 for _ in range(11)], []
max_score_diff = 0

def solution(n, info):
    global apeach_arrows, ans
    apeach_arrows = info
    
    dfs(0, n, [0 for __ in range(11)])
    print(max_score_diff)
    
    return ans if max_score_diff > 0 else [-1]


def dfs(depth, left_arrows, ryan_arrows):
    if depth == 11 or left_arrows == 0:
        global max_score_diff, ans
        ryan_arrows[10] += left_arrows
        
        cur_score_diff = calaulate_score_diff(ryan_arrows)
        if (max_score_diff < cur_score_diff or (max_score_diff == cur_score_diff and is_cur_ryan_smaller(ryan_arrows))):
            max_score_diff = cur_score_diff
            # ans = ryan_arrows
            ans = deepcopy(ryan_arrows)
            # print(max_score_diff, ans)
        
        ryan_arrows[10] -= left_arrows
        return
    
    dfs(depth + 1, left_arrows, ryan_arrows)
    if left_arrows > apeach_arrows[depth]:
        tmp = apeach_arrows[depth] + 1
        ryan_arrows[depth] = tmp
        dfs(depth + 1, left_arrows - tmp, ryan_arrows)
        ryan_arrows[depth] = 0
        
    
def calaulate_score_diff(ryan):
    score_diff = 0
    for i in range(11):
        if ryan[i] == 0 and apeach_arrows[i] == 0:
            continue
        score_diff += 10 - i if ryan[i] > apeach_arrows[i] else i - 10
    return score_diff


def is_cur_ryan_smaller(ryan):
    for i in range(10, -1, -1):
        if ryan[i] > ans[i]:
            return True
        elif ryan[i] < ans[i]:
            return False

File: 프로그래머스/lv2/92342. 양궁대회/양궁대회.java

class Solution {
	int[] apeachInfo;
	int size, maxScoreDiff = 0;
	int[] ans, wrongAns = {-1};
    public int[] solution(int n, int[] info) {
		apeachInfo = info.clone();
		size = n;

		int[] a = new int[11];
		ans = new int[11];
		for (int i = 0; i < 11; i++) {
			a[i] = 0;
			ans[i] = 0;
		}

		dfs(0, a, n);

        return (maxScoreDiff > 0) ? ans : wrongAns;
    }

	void dfs(int depth, int[] arrows, int leftArrows) {
		if (depth == 11 || leftArrows == 0) {
			arrows[10] += leftArrows;
			int curScoreDiff = calculateScore(arrows);

			if (curScoreDiff > maxScoreDiff || (curScoreDiff == maxScoreDiff && isMoreSmaller(arrows))) {
				maxScoreDiff = curScoreDiff;
				ans = arrows.clone();
			}

			arrows[10] -= leftArrows;
			return;
		}

		dfs(depth + 1, arrows, leftArrows);
		if (leftArrows > apeachInfo[depth]){
			int tmp = apeachInfo[depth] + 1;
			arrows[depth] = tmp;
			dfs(depth + 1, arrows, leftArrows - tmp);
			arrows[depth] = 0;
		}
	}

	int calculateScore(int[] ryan) {
		int score = 0;
        // int ryanScore = 0, apeachScore = 0;
		for (int i = 0; i < 11; i++) {
			if (ryan[i] == 0 && apeachInfo[i] == 0) continue;
            score += (ryan[i] > apeachInfo[i]) ? 10 - i : i - 10;
			// if (ryan[i] > apeachInfo[i]) 	ryanScore += 10 - i;
			// else 							apeachScore += 10 - i;
		}
		return score;
	}

	boolean isMoreSmaller(int[] ryan) {
		for (int i = 10; i > -1; i--) {
			if (ryan[i] == 0 && ans[i] == 0) continue;
			return ryan[i] > ans[i];
		}

		return true;
	}
}